File: modules/react-native-mlkit-face-detection/ios/RNMLKitFaceDetectionModule.swift

import ExpoModulesCore
import MLKitFaceDetection
import RNMLKitCore

// Function to reject a promise with a specified message and domain
func rejectPromiseWithMessage(promise: Promise, message: String, domain: String) {
    promise.reject(
        NSError(domain: domain, code: 1, userInfo: [NSLocalizedDescriptionKey: message])
    )
}

let ERROR_DOMAIN: String = "red.infinite.reactnativemlkit.FaceDetectorErrorDomain"


public class RNMLKitFaceDetectionModule: Module {
    var faceDetector: RNMLKitFaceDetector? = nil

    public func definition() -> ModuleDefinition {
        Name("RNMLKitFaceDetection")

        AsyncFunction("initialize") { (record: RNMLKitFaceDetectorOptionsRecord, promise: Promise) in
            let options = RNMLKitFaceDetectorOptions(record: record)
            self.faceDetector = RNMLKitFaceDetector(options: options)
            promise.resolve()
        }

        AsyncFunction("detectFaces") { (imagePath: String, promise: Promise) in
            let logger = Logger(logHandlers: [createOSLogHandler(category: Logger.EXPO_LOG_CATEGORY)])

            Task {
                do {
                    guard let faceDetector = self.faceDetector else {
                        rejectPromiseWithMessage(promise: promise, message: "[RNMLKitFaceDetection.detectFace] Face detector not initialized", domain: ERROR_DOMAIN)
                        return
                    }
                    let image = try RNMLKitImage(imagePath: imagePath)
                    let result = try await faceDetector.detectFaces(image: image)
                    logger.debug(result)
                    // Use result to resolve promise
                    promise.resolve(result.record)

                } catch {
                    rejectPromiseWithMessage(promise: promise, message: "[RNMLKitFaceDetection.detectFace] Error Detecting Faces: \(error)", domain: ERROR_DOMAIN)
                }
            }
        }
    }
}

File: modules/react-native-mlkit-face-detection/android/src/main/java/red/infinite/reactnativemlkit/facedetection/RNMLKitFaceDetectionModule.kt
package red.infinite.reactnativemlkit.facedetection

import RNMLKitFaceDetector
import RNMLKitFaceDetectorOptions
import RNMLKitFaceDetectorOptionsRecord
import android.net.Uri

import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.runBlocking
import red.infinite.reactnativemlkit.core.RNMLKitModule


class RNMLKitFaceDetectionModule : RNMLKitModule("RNMLKitFaceDetection") {
    private var faceDetector: RNMLKitFaceDetector? = null

    // Each module class must implement the definition function. The definition consists of components
    // that describes the module's functionality and behavior.
    // See https://docs.expo.dev/modules/module-api for more details about available components.
    override fun definition() = ModuleDefinition {
        // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
        // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
        // The module will be accessible from `requireNativeModule('RNMLKitFaceDetection')` in JavaScript.
        Name("RNMLKitFaceDetection")

        AsyncFunction("initialize") { options: RNMLKitFaceDetectorOptionsRecord, promise: Promise ->
            log.d("initialize: Initializing Face Detection")
            faceDetector = RNMLKitFaceDetector(RNMLKitFaceDetectorOptions(options))
            log.d("initialize: Initializing Face Detection")
            promise.resolve(null)
        }

        AsyncFunction("detectFaces") { imagePath: String, promise: Promise ->

            val assetManager = context.assets

            assetManager.list("/")?.mapNotNull { asset ->
                log.d(asset)
            }

            val imageUri = Uri.parse(imagePath)

            runBlocking suspend@{
                val faceDetector = faceDetector
                if (faceDetector == null) {
                    handleException(promise, message = "Face Detector not initialized")
                    return@suspend
                }
                try {
                    var result =
                        faceDetector.detectFaces(imagePath, appContext).getOrElse {
                            handleException(promise, it)
                            return@suspend
                        }
                    promise.resolve(result.record)
                } catch (e: Exception) {
                    handleException(promise, e)
                }

            }
        }
    }
}
